### DISCUSSION TITLE
How to cache results from API calls when building SSG pages? Traditional in-memory memoization doesn't seem to work

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/next.js#13765] [Help] [answered]
I'm fetching the Airtable API 3 times per page, and I'm building around ~30 pages, which generates about 100 API calls when building the site on Vercel.

```
17:58:38.370  
Fetching airtable API at "https://api.airtable.com/v0/app25whOKJt4iOE1z/Product"
17:58:38.400  
[Product] 3 airtable API records fetched
17:58:38.527  
Cache key does not exist
17:58:38.527  
{}
17:58:38.527  
Fetching airtable API at "https://api.airtable.com/v0/app25whOKJt4iOE1z/Customer"
17:58:38.529  
Cache key does not exist
17:58:38.529  
{}
17:58:38.529  
Fetching airtable API at "https://api.airtable.com/v0/app25whOKJt4iOE1z/Customer"
17:58:38.687  
[Product] 3 airtable API records fetched
17:58:38.773  
[Customer] 2 airtable API records fetched
17:58:38.773  
Cache key does not exist
17:58:38.773  
{
17:58:38.773  
  CustomerTable: { timestamp: 1591286318773, value: { records: [Array] } }
17:58:38.773  
}
17:58:38.773  
Fetching airtable API at "https://api.airtable.com/v0/app25whOKJt4iOE1z/Theme"
17:58:38.884  
Cache key does not exist
17:58:38.884  
{}
17:58:38.884  
Fetching airtable API at "https://api.airtable.com/v0/app25whOKJt4iOE1z/Customer"
17:58:38.926  
[Customer] 2 airtable API records fetched
17:58:38.926  
Cache key does not exist
17:58:38.927  
{
17:58:38.927  
  CustomerTable: { timestamp: 1591286318926, value: { records: [Array] } }
17:58:38.927  
}
```

I tried implementing a basic in-memory cache system (memoization), but while it does work locally (unit tested), it doesn't work when the site is being build.

Here is the cache implementation:
https://github.com/UnlyEd/next-right-now/pull/92

It relies on treating the file `memoizeWithTTLCache` as a node file being loaded in memory as a singleton.

But it obviously don't work neither on Vercel, neither locally when running `next build`.

Am I missing something? I believe pages are built in a special way that resets the cache between each page. (as logged above, the cache seems to be filled with `CustomerTable`, then with `ThemeTable`, but then the next page gets an empty cache and starts again)

**Comment by lfades** [comment, marked as answer]
We may need an example to show how to do this 🤔 

You could create a build time only file that stores the data and read from that file instead of repeating the request, we do something like that for our docs. In our docs we fetch the latest tag of the repo to get the manifest of routes, and because that can happen once per documentation page, we cache the result in a file (build time only) to avoid being rate limited:

```js
import path from 'path';
import { readFile, writeFile } from '../fs-utils';
import { GITHUB_API_URL, REPO_NAME } from './constants';

const USE_CACHE = process.env.USE_CACHE === 'true';
const TAG_CACHE_PATH = path.resolve('.github-latest-tag');

export async function getLatestTag() {
  let cachedTag;

  if (USE_CACHE) {
    try {
      cachedTag = await readFile(TAG_CACHE_PATH, 'utf8');
    } catch (error) {
      // A cached file is not required
    }
  }

  if (!cachedTag) {
    const res = await fetch(`${GITHUB_API_URL}/repos/${REPO_NAME}/releases/latest`);

    if (res.ok) {
      const data = await res.json();
      const tag = data.tag_name;

      if (USE_CACHE) {
        try {
          await writeFile(TAG_CACHE_PATH, tag, 'utf8');
        } catch (error) {
          // A cached file is not required
        }
      }

      cachedTag = tag;
    }
  }

  return cachedTag;
}
```

(You can also [check the above file in our repo](https://github.com/vercel/next-site/blob/e09472707419213a6b26fdcd788ef53e1fe17853/lib/github/api.js))

We then set `process.env.USE_CACHE` inside `next.config.js`:

```js
module.exports = {
  webpack: (config, { dev, isServer }) => {
    if (!dev && isServer) {
      // we're in build mode so enable shared caching for the GitHub API
      process.env.USE_CACHE = 'true';
    }

    return config;
  }
}
```

(You can also [check the above file in our repo](https://github.com/vercel/next-site/blob/e09472707419213a6b26fdcd788ef53e1fe17853/next.config.js))

And include the generated cache file (`.github-latest-tag`) to `.gitignore` to avoid pushing it to the repo accidentally.

**Reply by Vadorequest** [reply]
Thank you very much. This is very helpful.

I could see the need for a NPM package to deal with this. (e.g: `@next/cache-pages`)

Also beware your above implementation will silent errors when `!res.ok` (fetch errors, etc.)

Do you recommend writing such tmp files in the root directory? I'm thinking about a `.build-cache` folder (probably don't change a thing).

Also, I'm wondering about a more generic variable name for `process.env.USE_CACHE = 'true';`. 
I'm thinking about `NEXT_BUILD_MODE = 'initialBuild'`, but I'm not sure about what other modes could be useful to developers (I'm developing a Next.js boilerplate, so that kind of consideration matters)

In my case, I'd rather set a `NEXT_BUILD_MODE` env var, and combine it with `process.env.NODE_ENV` at runtime to know if I should apply the caching strategy.
Additionally, I could use this `NEXT_BUILD_MODE` to print additional logging (or silent it!), but I'm not sure about what other values could be stored within (e.g: when `isServer is false`) because I don't quite understand what it means.

if you have any idea/feedback about that, that'd be awesome :)

**Reply by lfades** [reply]
Error handling can be improved there. This example goes to our site and it works well there, if it doesn't return a tag our docs fail at build time. A public implementation won't be this simple so feel free to improve it 👍 .

Keep in mind this is only intended for build time, this caching is not intended to be used at runtime where the environment may not even allow you to write or access shared cache files.

**Reply by Vadorequest** [reply]
But, if I do that on Vercel, will it work? At build time, obviously (but it's still using the file system, AFAIK)

**Reply by lfades** [reply]
At build time yes, our docs deploy to Vercel 👍

**Reply by Vadorequest** [reply]
@lfades Thanks again for your help, I managed to write a re-usable hybrid memory/disk cache storage for this.

Went from 140 API calls to half a dozen:

![image](https://user-images.githubusercontent.com/3807458/83927233-9a0c9f80-a78c-11ea-8810-683d67383077.png)

---

**Implementation:**

Check the NRN repository below to see the git tree at time of writing:
https://github.com/UnlyEd/next-right-now/blob/e9d07ff788/src/utils/caching/hybridCache.ts

Check PR:
https://github.com/UnlyEd/next-right-now/pull/92

```ts
/**
 * Hybrid cache, can use in-memory storage or on-disk storage.
 *
 * You can select the kind of storage you want through the options.
 *
 * Algorithm:
 *  - Check if cache exist for key, if not then call the dataResolver and cache the result.
 *  - If cache exists and record age is lower than TTL (default: 30 seconds), then return cached data. Otherwise, call the dataResolver and cache the result.
 *
 * Special considerations:
 *  - If options.ttl is 0, then TTL is disabled and data will live indefinitely
 *  - Memory cache storage is selected by default
 *  - Memory cache storage will not work if you want to cache data at build time, when building the pages (typically: SSG). You must use the disk for this kind of usage.
 *  - Disk cache storage works on Vercel, you need a writable file system
 *
 * @example With default (in-memory, TTL 30 secs)
 *  const stuff = await hybridCache('CustomerTable', async () => await fetchMyStuff());
 *
 * @example Using disk storage with no TTL, meant to cache stuff once for all pages during server initial build (for SSG, typically - IS_SERVER_INITIAL_BUILD comes from next.config.js)
 *  const stuff = await hybridCache('CustomerTable', async () => await fetchMyStuff(), { enabled: !!process.env.IS_SERVER_INITIAL_BUILD && process.env.NODE_ENV !== 'development', storage: {type: 'disk', options: {filename: 'my-stuff'}} });
 *
 * @param keyResolver string | Function Key used to store the data in the cache
 * @param dataResolver Function Has to be implemented by the caller and is supposed to take care of fetching the data
 * @param options HybridCacheOptions
 */
const hybridCache = async <T>(keyResolver: string | (() => string), dataResolver: () => T, options: Partial<HybridCacheOptions> = defaultHybridCacheOptions): Promise<T> => {
  const { ttl, enabled, storage } = deepmerge(defaultHybridCacheOptions, options);

  if (!enabled) { // Bypasses cache completely
    // eslint-disable-next-line no-console
    console.debug('Cache is disabled, bypassing');
    return dataResolver();
  }
  let cacheStorage: HybridCacheStorage;
  let storageOptions: StorageOptions = {};

  if (storage.type === 'memory') {
    cacheStorage = require('./memoryCacheStorage');
  } else {
    cacheStorage = require('./diskCacheStorage');
    const { options } = storage;
    storageOptions = options;
  }

  let key: string;

  if (typeof keyResolver === 'function') {
    key = keyResolver();
  } else {
    key = keyResolver;
  }

  const cachedItem: CachedItem = await cacheStorage.get(key, storageOptions);

  if (typeof cachedItem !== 'undefined') {
    const { value, timestamp } = cachedItem;

    if (timestamp === 0 || getTimestampsElapsedTime(timestamp, +new Date()) < ttl) {
      return value;
    } else {
      logger.debug('Cache key has expired');
    }
  } else {
    logger.debug('Cache key does not exist');
  }

  const unMemoizedResult: T = await dataResolver();
  return cacheStorage.set(key, unMemoizedResult, storageOptions);
};
```

**Reply by pip8786** [reply]
@lfades Thanks for your example, I was able to build a caching system for my app. Just wanted to point out that your links are broken and I tried looking around for the files but couldn't find anything like next-site in your repos.

**Comment by nk9** [comment]
The Next.js docs [now describe a solution](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-third-party-libraries) to this problem using React's `cache` function. However, `cache` [presently has two caveats](https://react.dev/reference/react/cache) which aren't mentioned in the Next docs:

- `cache` is only for use with [React Server Components](https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#react-server-components).
- `cache` is only available in React’s [Canary](https://react.dev/community/versioning-policy#canary-channel) and [experimental](https://react.dev/community/versioning-policy#experimental-channel) channels.

So does Next.js presently recommend using a React Canary release and `cache`? Also, I don't know anything about React Server Components. Is code in server-side Next.js functions like `getStaticProps` run in a Server Component, and eligible for use with `cache`?

**Comment by nwatab** [comment]
Caching Internet data into disk is shown above. Then, how to cache data on disk into memory?
